apps/heft/src/plugins/CopyFilesPlugin.ts

```diff
@@ -0,0 +1,336 @@
+      buildFolder,
+      copyConfigurations
+    );
+
+    if (copyDescriptors.length === 0) {
+      // No need to run copy and print to console
+      return;
+    }
+
+    const { copiedFileCount, linkedFileCount } = await this.copyFilesAsync(copyDescriptors);
+    const duration: number = performance.now() - startTime;
+    logger.terminal.writeLine(
+      `Copied ${copiedFileCount} file${copiedFileCount === 1 ? '' : 's'} and ` +
+        `linked ${linkedFileCount} file${linkedFileCount === 1 ? '' : 's'} in ${Math.round(duration)}ms`
+    );
+
+    // Then enter watch mode if requested
+    if (options.watchMode) {
+      await this._runWatchAsync(options);
+    }
+  }
+
+  protected async copyFilesAsync(copyDescriptors: ICopyFileDescriptor[]): Promise<ICopyFilesResult> {
+    if (copyDescriptors.length === 0) {
+      return { copiedFileCount: 0, linkedFileCount: 0 };
+    }
+
+    let copiedFileCount: number = 0;
+    let linkedFileCount: number = 0;
+    await Async.forEachLimitAsync(
+      copyDescriptors,
+      Constants.maxParallelism,
+      async (copyDescriptor: ICopyFileDescriptor) => {
+        if (copyDescriptor.hardlink) {
+          const hardlinkPromises: Promise<void>[] = copyDescriptor.destinationFilePaths.map(
+            (destinationFilePath) => {
+              return FileSystem.createHardLinkAsync({
+                linkTargetPath: copyDescriptor.sourceFilePath,
+                newLinkPath: destinationFilePath,
+                alreadyExistsBehavior: AlreadyExistsBehavior.Overwrite
+              });
+            }
+          );
+          await Promise.all(hardlinkPromises);
+
+          linkedFileCount++;
+        } else {
+          // If it's a copy, we will call the copy function
+          if (copyDescriptor.destinationFilePaths.length === 1) {
+            await FileSystem.copyFileAsync({
+              sourcePath: copyDescriptor.sourceFilePath,
+              destinationPath: copyDescriptor.destinationFilePaths[0],
+              alreadyExistsBehavior: AlreadyExistsBehavior.Overwrite
+            });
+          } else {
+            await FileSystem.copyFileToManyAsync({
+              sourcePath: copyDescriptor.sourceFilePath,
+              destinationPaths: copyDescriptor.destinationFilePaths,
+              alreadyExistsBehavior: AlreadyExistsBehavior.Overwrite
+            });
+          }
+
+          copiedFileCount++;
+        }
+      }
+    );
+
+    return {
+      copiedFileCount,
+      linkedFileCount
+    };
+  }
+
+  private async _getCopyFileDescriptorsAsync(
+    buildFolder: string,
+    copyConfigurations: IExtendedSharedCopyConfiguration[]
+  ): Promise<ICopyFileDescriptor[]> {
+    // Create a map to deduplicate and prevent double-writes
+    const destinationCopyDescriptors: Map<string, ICopyFileDescriptor> = new Map();
+    // And a map to contain the actual results
+    const sourceCopyDescriptors: Map<string, ICopyFileDescriptor> = new Map();
```

Comment:
For `destinationCopyDescriptors` and `sourceCopyDescriptors` document what the `Map` key is.  Absolute path to a source file?